transactions/demo-rust/src/lib.rs

```diff
@@ -0,0 +1,208 @@
+            csl::tx_builder::CoinSelectionStrategyCIP2::RandomImproveMultiAsset,
+        )
+        .unwrap();
+    tx_builder.add_required_signer(own_pkh);
+
+    tx_builder
+}
+
+/// Make a transaction that releases UTxO stored at the `EqValidator` in one of the cases below:
+/// - redeemer is `IsEqual` and the supplied plutus data is equal to the one locked as datum
+/// - redeemer is `IsNotEqual` and the supplied plutus data is not equal to the one locked as datum
+///
+/// This is a pure function to build a transaction scaffold.
+pub fn mk_claim_tx(
+    own_pkh: &Ed25519KeyHash,
+    own_addr: &csl::address::Address,
+    own_utxos: &BTreeMap<csl::TransactionInput, csl::TransactionOutput>,
+    eq_validator: &csl::plutus::PlutusScript,
+    eq_validator_utxos: &BTreeMap<csl::TransactionInput, csl::TransactionOutput>,
+    tx_input: &csl::TransactionInput,
+    collateral: &csl::TransactionInput,
+    eq_redeemer: &EqRedeemer,
+) -> csl::tx_builder::TransactionBuilder {
+    let mut tx_builder = create_tx_builder();
+    let tx_input_resolved = eq_validator_utxos.get(tx_input).unwrap();
+    let value = tx_input_resolved.amount();
+    let datum = tx_input_resolved
+        .plutus_data()
+        .expect("Expected an inline datum in transaction input.");
+
+    let redeemer = to_redeemer(&convert_plutus_data(eq_redeemer.to_plutus_data()));
+    let mut tx_inputs_builder = TxInputsBuilder::new();
+    let data_cost = csl::DataCost::new_coins_per_byte(&csl::utils::to_bignum(COINS_PER_UTXO_WORD));
+
+    let tx_input_witness = PlutusWitness::new(eq_validator, &datum, &redeemer);
+    tx_inputs_builder.add_plutus_script_input(&tx_input_witness, tx_input, &value);
+
+    let mut available_inputs = csl::utils::TransactionUnspentOutputs::new();
+    own_utxos.iter().for_each(|(tx_in, tx_out)| {
+        let utxo = csl::utils::TransactionUnspentOutput::new(&tx_in, &tx_out);
+        available_inputs.add(&utxo);
+    });
+
+    let mut collateral_builder = TxInputsBuilder::new();
+    let collateral_amount = own_utxos.get(&collateral).unwrap().amount();
+    collateral_builder.add_input(&own_addr, &collateral, &collateral_amount);
+
+    tx_builder
+        .add_output(
+            &TransactionOutputBuilder::new()
+                .with_address(&own_addr)
+                .next()
+                .unwrap()
+                .with_asset_and_min_required_coin_by_utxo_cost(&csl::MultiAsset::new(), &data_cost)
+                .unwrap()
+                .build()
+                .unwrap(),
+        )
+        .unwrap();
+
+    tx_builder.set_inputs(&tx_inputs_builder);
+    tx_builder.set_collateral(&collateral_builder);
+    tx_builder
+        .add_inputs_from(
+            &available_inputs,
+            csl::tx_builder::CoinSelectionStrategyCIP2::RandomImproveMultiAsset,
+        )
+        .unwrap();
+    tx_builder.add_required_signer(own_pkh);
+
+    tx_builder
+}
+
+/// Builds and submits `mk_lock_tx`
+pub async fn lock_tx_build_and_submit(
```

Comment:
I love that we have the pure part, and the IO part.

I want to refactor this into a separate module though, not sure where. Why?

Here we make a bunch of choices that are arbitrary, for example, Ogmios, that are very specific to offchain orchestration that can be quite varied.

I placed that under test, as I see it as a 'test application'. Maybe under `tests/transactions.rs`?


So like `src/transactions.rs` is pure transactions that is the core of the protocol api, and `tests/transactions.rs` is a particular orchestration?